release-charm/README.md

```diff
@@ -0,0 +1,64 @@
+# canonical/charming-actions/release-charm
+
+This action is used to release an already uploaded charm to a different channel in charmhub. It is designed to be manually triggered with the inputs . The revision of the charm must be uploaded using `upload-charm` action. It updates the existing revision release tag with the destination channel and timestamp.
+
+## Usage
+
+```yaml
+name: Release charm to other tracks and channels
+
+on:
+  workflow_dispatch:
+    inputs:
+      destination-channel:
+        description: 'Destination Channel'
+        required: true
+      origin-channel:
+        description: 'Origin Channel'
+        required: false
+      rev:
+        description: 'Revision number'
+        required: false
+      tag-prefix:
+        description: 'Tag Prefix'
+        required: false
+      charm-path:
+        description: 'Path to charm'
+        required: false
```

Comment:
Why are these needed? We should only promote what is already published in some other channel. To me, it makes sense to keep the workflow strict and not allow too much deviation or customizability.

Reply:
This action was designed to be triggered manually with certain input. I think it would be helpful to include the yaml for inputs as reference. 
The key inputs are destination channel and origin channel/revision. I don't know a way to require either origin channel or revision. That's why I left both as not required.
tag-prefix and charm-path are for repos with multiple charm. You could omit those fields for single charm repos.

Reply:
I had some similar comments over [here](https://github.com/canonical/istio-operators/pull/80/files/11b49e2ffa3145b2658d461222fb2214f3f991c8#r844119799).  Summary: if rev number and origin channel are mutually exclusive, we should highlight that in their descriptions and the actual code should error out if it receives both (otherwise people will accidentally specify both and not know what result they'll receive).

Reply:
But more fundamentally, @simskij I see you're wondering why we need `revision number` at all.  I think this is needed for the case where something is not yet in a channel at all.

Reply:
@ca-scribner @agathanatasha I don't think we should allow publishing something not already in a channel. upload-charm will put it on edge, and then that revision should be what gets promoted. Opening up for skipping over channels is a disaster waiting to happen.

Reply:
@simskij I disagree.  Can you give an example of the disaster you're thinking of?  I'm having trouble thinking of a scenario where putting a charm from unpublished to `candidate`, etc., has a bad consequence

The example that comes to mind for why this is needed is:
1. a new revision of the charm is created and put into charmhub (`rev1`) on `latest/edge`
2. we start doing QA to see if `rev1` is good for release (ask field, push to another team member, etc)
3. while doing QA, someone finishes a new feature not intended for a near-term release and pushes it to main.  This triggers CI to push `rev2` to `latest/edge`
4. QA process passes and we want to release `rev1` to `latest/stable`

How can we complete the release in step (4)?

Reply:
A bit removed from this conversation, but I find it off that are dividing our release CI building blocks into:
* upload-charm: build, upload to charmhub, release to edge
* release: release to track/[^edge]

By convention it totally makes sense that our publish CI build, upload, and release to edge because we decided that `latest/edge` should track with `main`.  But those are two basic actions, not one.  I don't think that means releasing to edge should be tightly coupled with build/upload.  I think the name of the action alone is a good guide - `upload-charm` does not suggest anything is being released, and I can't think of another natural name that doesn't have "and" in it (eg: `upload-and-release-charm`)

This makes more sense to me:

upload-charm: build, upload to charmhub
release-charm: release to anywhere (either from unreleased to a channel, or from one channel to another)

I like that our publish CI uploads and releases to `latest/edge` so I propose we keep that, just that it shouldn't be one action.  publish CI could just do both